dbos/system_database.py

```diff
@@ -830,3 +869,68 @@ def get_event(
                 }
             )
         return value
+
+    def _flush_workflow_status_buffer(self) -> None:
+        """
+        Export the workflow status buffer to the database, up to the batch size.
+        """
+        if len(self._workflow_status_buffer) == 0:
+            return
+
+        with self.engine.begin() as c:
+            exported = 0
+            local_batch_size = min(
+                buffer_flush_batch_size, len(self._workflow_status_buffer)
+            )
+            while exported < local_batch_size:
+                # Pop the first key in the buffer (FIFO)
+                wf_id = next(iter(self._workflow_status_buffer))
+                status = self._workflow_status_buffer.pop(wf_id)
+                self.update_workflow_status(status, conn=c)
+                exported += 1
+
+    def _flush_workflow_inputs_buffer(self) -> None:
+        """
+        Export the workflow inputs buffer to the database, up to the batch size.
+        """
+        if len(self._workflow_inputs_buffer) == 0:
+            return
+
+        with self.engine.begin() as c:
+            exported = 0
+            local_batch_size = min(
+                buffer_flush_batch_size, len(self._workflow_inputs_buffer)
+            )
+            while exported < local_batch_size:
+                wf_id = next(iter(self._workflow_inputs_buffer))
+                inputs = self._workflow_inputs_buffer.pop(wf_id)
+                self.update_workflow_inputs(wf_id, inputs, conn=c)
+                exported += 1
+
+    def flush_workflow_buffers(self) -> None:
```

Comment:
Needs some way to recover from exceptions--if the DB is overloaded at all, this will die forever

Reply:
Done.